core/src/main/java/hudson/model/Queue.java

```diff
@@ -3009,4 +3011,70 @@ public static Queue getInstance() {
     public static void init(Jenkins h) {
         h.getQueue().load();
     }
+
+    /**
+     * Schedule <tt>Queue.save()</tt> call for near future once items change. Ignore all changes until the time the save
+     * takes place.
+     *
+     * Once queue is restored after a crash, items stages might not be accurate until the next #maintain() - this is not
+     * a problem as the items will be reshuffled first and then scheduled during the next maintainance cycle.
+     *
+     * Implementation note: Queue.load() calls QueueListener hooks for every item deserialized that can hammer the persistance
+     * on load. The problem is avoided by delaying the actual save for the time long enough for queue to load so the save
+     * operations will collapse into one. Also, items are persisted as buildable or blocked in vast majority of cases and
+     * those stages does not trigger the save here.
+     */
+    @Extension
+    @Restricted(NoExternalUse.class)
+    public static final class Saver extends QueueListener implements Runnable {
+        private static final int DISABLED = -1;
+        private static final int MIN_DELAY = 10;
```

Comment:
10s is going to be much to long for some users on busy systems that really *can't* afford to lose queue info - maybe more like 1s?  That's often enough not to have Queue write-thrashing.

Reply:
Well, to be sure this is reasonable safe we would have to benchmark against extremely loaded system on a super slow filesystem. My thinking is, Jenkins never provided such guarantees and now we are providing some. If someone need realtime guarantee, we can always relax the limits later provided it will be demonstrated it is safe...

Reply:
@olivergondza But they'd have to explicitly reduce the default to trigger any issue, right?  Seems like if you go out of your way to change a System Property and it causes problems, that's not Jenkins' problem. 

Also, saving the Queue *should* be pretty fast on most systems (like 100ms or less).

Reply:
Alright, let's keep this simple and let users to configure what they want. We can turn to being extra protective if/when people would complain this let them to shoot themselves in the leg way too easily.

Reply:
@svanoort serialising items/writing to file is not memory friendly procedure - so, it could be fast, but still too expensive.

Reply:
@Jimilian Yeah, it can generate a decent amount of memory garbage.  A much larger chunk of garbage I see is due to short-lived buffers (either explicit or primitive arrays) created in I/O though.  If you use [my optimized GC settings](https://jenkins.io/blog/2016/11/21/gc-tuning/) it should be able to clear ~1 GB/s per CPU core participating in GC. 

I think for the reasons you cited a long default is reasonable, but we should let users control this to tune for their cases.